File: aez_amd64.py
#!/usr/bin/env python3
#
# To the extent possible under law, Yawning Angel has waived all copyright
# and related or neighboring rights to aez, using the Creative
# Commons "CC0" public domain dedication. See LICENSE or
# <http://creativecommons.org/publicdomain/zero/1.0/> for full details.

#
# Dependencies: https://github.com/Maratyszcza/PeachPy
#
# python3 -m peachpy.x86_64 -mabi=goasm -S -o aez_amd64.s aez_amd64.py
#

from peachpy import *
from peachpy.x86_64 import *

cpuidParams = Argument(ptr(uint32_t))

with Function("cpuidAMD64", (cpuidParams,)):
    reg_params = registers.r15
    LOAD.ARGUMENT(reg_params, cpuidParams)

    MOV(registers.eax, [reg_params])
    MOV(registers.ecx, [reg_params+4])

    CPUID()

    MOV([reg_params], registers.eax)
    MOV([reg_params+4], registers.ebx)
    MOV([reg_params+8], registers.ecx)
    MOV([reg_params+12], registers.edx)

    RETURN()

a = Argument(ptr(const_uint8_t))
b = Argument(ptr(const_uint8_t))
c = Argument(ptr(const_uint8_t))
d = Argument(ptr(const_uint8_t))
dst = Argument(ptr(uint8_t))

with Function("xorBytes1x16AMD64SSE2", (a, b, dst)):
    reg_a = GeneralPurposeRegister64()
    reg_b = GeneralPurposeRegister64()
    reg_dst = GeneralPurposeRegister64()

    LOAD.ARGUMENT(reg_a, a)
    LOAD.ARGUMENT(reg_b, b)
    LOAD.ARGUMENT(reg_dst, dst)

    xmm_a = XMMRegister()
    xmm_b = XMMRegister()

    MOVDQU(xmm_a, [reg_a])
    MOVDQU(xmm_b, [reg_b])

    PXOR(xmm_a, xmm_b)

    MOVDQU([reg_dst], xmm_a)

    RETURN()

with Function("xorBytes3x16AMD64SSE2", (a, b, c, dst)):
    reg_a = GeneralPurposeRegister64()
    reg_b = GeneralPurposeRegister64()
    reg_c = GeneralPurposeRegister64()
    reg_dst = GeneralPurposeRegister64()

    LOAD.ARGUMENT(reg_a, a)
    LOAD.ARGUMENT(reg_b, b)
    LOAD.ARGUMENT(reg_c, c)
    LOAD.ARGUMENT(reg_dst, dst)

    xmm_a = XMMRegister()
    xmm_b = XMMRegister()
    xmm_c = XMMRegister()

    MOVDQU(xmm_a, [reg_a])
    MOVDQU(xmm_b, [reg_b])
    MOVDQU(xmm_c, [reg_c])

    PXOR(xmm_a, xmm_b)
    PXOR(xmm_a, xmm_c)

    MOVDQU([reg_dst], xmm_a)

    RETURN()

with Function("xorBytes4x16AMD64SSE2", (a, b, c, d, dst)):
    reg_a = GeneralPurposeRegister64()
    reg_b = GeneralPurposeRegister64()
    reg_c = GeneralPurposeRegister64()
    reg_d = GeneralPurposeRegister64()
    reg_dst = GeneralPurposeRegister64()

    LOAD.ARGUMENT(reg_a, a)
    LOAD.ARGUMENT(reg_b, b)
    LOAD.ARGUMENT(reg_c, c)
    LOAD.ARGUMENT(reg_d, d)
    LOAD.ARGUMENT(reg_dst, dst)

    xmm_a = XMMRegister()
    xmm_b = XMMRegister()
    xmm_c = XMMRegister()
    xmm_d = XMMRegister()

    MOVDQU(xmm_a, [reg_a])
    MOVDQU(xmm_b, [reg_b])
    MOVDQU(xmm_c, [reg_c])
    MOVDQU(xmm_d, [reg_d])

    PXOR(xmm_a, xmm_b)
    PXOR(xmm_c, xmm_d)
    PXOR(xmm_a, xmm_c)

    MOVDQU([reg_dst], xmm_a)

    RETURN()

#
# Sigh.  PeachPy has "interesting" ideas of definitions for certain things,
# so just use the `zen` uarch, because it supports everything.
#

s = Argument(ptr(uint8_t))
k = Argument(ptr(const_uint8_t))

with Function("aes4AMD64AESNI", (s, k), target=uarch.zen):
    reg_s = GeneralPurposeRegister64()
    reg_k = GeneralPurposeRegister64()

    LOAD.ARGUMENT(reg_s, s)
    LOAD.ARGUMENT(reg_k, k)

    xmm_state = XMMRegister()
    xmm_i = XMMRegister()
    xmm_j = XMMRegister()
    xmm_l = XMMRegister()
    xmm_zero = XMMRegister()

    PXOR(xmm_zero, xmm_zero)
    MOVDQU(xmm_state, [reg_s])
    MOVDQU(xmm_i, [reg_k])
    MOVDQU(xmm_j, [reg_k+16])
    MOVDQU(xmm_l, [reg_k+32])

    AESENC(xmm_state, xmm_j)
    AESENC(xmm_state, xmm_i)
    AESENC(xmm_state, xmm_l)
    AESENC(xmm_state, xmm_zero)

    MOVDQU([reg_s], xmm_state)

    RETURN()

with Function("aes10AMD64AESNI", (s, k), target=uarch.zen):
    reg_s = GeneralPurposeRegister64()
    reg_k = GeneralPurposeRegister64()

    LOAD.ARGUMENT(reg_s, s)
    LOAD.ARGUMENT(reg_k, k)

    xmm_state = XMMRegister()
    xmm_i = XMMRegister()
    xmm_j = XMMRegister()
    xmm_l = XMMRegister()
    xmm_zero = XMMRegister()

    PXOR(xmm_zero, xmm_zero)
    MOVDQU(xmm_state, [reg_s])
    MOVDQU(xmm_i, [reg_k])
    MOVDQU(xmm_j, [reg_k+16])
    MOVDQU(xmm_l, [reg_k+32])

    AESENC(xmm_state, xmm_i)
    AESENC(xmm_state, xmm_j)
    AESENC(xmm_state, xmm_l)
    AESENC(xmm_state, xmm_i)
    AESENC(xmm_state, xmm_j)
    AESENC(xmm_state, xmm_l)
    AESENC(xmm_state, xmm_i)
    AESENC(xmm_state, xmm_j)
    AESENC(xmm_state, xmm_l)
    AESENC(xmm_state, xmm_i)

    MOVDQU([reg_s], xmm_state)

    RETURN()

File: aez_amd64.go
// aez_amd64.go - AMD64 specific routines.
//
// To the extent possible under law, Yawning Angel has waived all copyright
// and related or neighboring rights to aez, using the Creative
// Commons "CC0" public domain dedication. See LICENSE or
// <http://creativecommons.org/publicdomain/zero/1.0/> for full details.

// +build amd64,!gccgo,!appengine

package aez

//go:noescape
func cpuidAMD64(cpuidParams *uint32)

//go:noescape
func xorBytes1x16AMD64SSE2(a, b, dst *byte)

//go:noescape
func xorBytes4x16AMD64SSE2(a, b, c, d, dst *byte)

//go:noescape
func aezE4AMD64AESNI(j, i, l, k, s, dst *byte)

//go:noescape
func aezE10AMD64AESNI(l, k, s, dst *byte)

func xorBytes1x16(a, b, dst []byte) {
	xorBytes1x16AMD64SSE2(&a[0], &b[0], &dst[0])
}

func xorBytes4x16(a, b, c, d []byte, dst *[blockSize]byte) {
	xorBytes4x16AMD64SSE2(&a[0], &b[0], &c[0], &d[0], &dst[0])
}

type roundAESNI struct {
	keys [extractedKeySize]byte
}

func (r *roundAESNI) Reset() {
	memwipe(r.keys[:])
}

func (r *roundAESNI) E4(j, i, l *[blockSize]byte, src []byte, dst *[blockSize]byte) {
	aezE4AMD64AESNI(&j[0], &i[0], &l[0], &r.keys[0], &src[0], &dst[0])
}

func (r *roundAESNI) E10(l *[blockSize]byte, src []byte, dst *[blockSize]byte) {
	aezE10AMD64AESNI(&l[0], &r.keys[0], &src[0], &dst[0])
}

func newRoundAESNI(extractedKey *[extractedKeySize]byte) aesImpl {
	r := new(roundAESNI)
	copy(r.keys[:], extractedKey[:])

	return r
}

func supportsAESNI() bool {
	const (
		aesniBit   = 1 << 25
		osXsaveBit = 1 << 27
	)

	// Check to see if the OS knows how to save/restore XMM state.
	// CPUID.(EAX=01H, ECX=0H):ECX.OSXSAVE[bit 27]==1
	regs := [4]uint32{0x01}
	cpuidAMD64(&regs[0])
	if regs[2]&osXsaveBit == 0 {
		return false
	}

	// Check for AES-NI support.
	// CPUID.(EAX=01H, ECX=0H):ECX.AESNI[bit 25] = 1
	return regs[2]&aesniBit != 0
}

func init() {
	if supportsAESNI() {
		newAes = newRoundAESNI
	}
}
